File: sdks/go/cvt/producer/producer.go
package producer

import (
	"bytes"
	"context"
	"encoding/json"
	"io"
	"log"
	"net/http"
	"sync"
)

// Producer handles server-side validation of HTTP requests and responses.
type Producer struct {
	config Config
	mu     sync.RWMutex
}

// NewProducer creates a new Producer with the given configuration.
func NewProducer(config Config) (*Producer, error) {
	// Apply defaults
	if config.Mode == "" {
		config.Mode = ModeStrict
	}
	if !config.ValidateRequest && !config.ValidateResponse {
		config.ValidateRequest = true
		config.ValidateResponse = true
	}

	if err := config.Validate(); err != nil {
		return nil, err
	}

	return &Producer{
		config: config,
	}, nil
}

// ValidateRequest validates an incoming HTTP request against the schema.
// Returns the validation result. For Strict mode, check result.Valid to decide
// whether to reject the request.
func (p *Producer) ValidateRequest(ctx context.Context, r *http.Request, body []byte) *ValidationResult {
	result := &ValidationResult{
		Type:  "request",
		Valid: true,
	}

	if !p.config.ValidateRequest {
		return result
	}

	// Build interaction for request-only validation
	interaction := &Interaction{
		Method:  r.Method,
		Path:    p.buildPath(r),
		Headers: p.extractHeaders(r.Header),
		Body:    string(body),
		// Use a minimal valid response to pass through - we're only validating request
		StatusCode:   200,
		ResponseBody: "{}",
	}

	validationResult, err := p.validateInteraction(ctx, interaction)
	if err != nil {
		result.Valid = false
		result.Errors = []string{err.Error()}
		return result
	}

	// Check if the result contains request-specific errors
	result.Valid = validationResult.Valid
	result.Errors = validationResult.Errors

	return result
}

// ValidateResponse validates an outgoing HTTP response against the schema.
func (p *Producer) ValidateResponse(ctx context.Context, r *http.Request, reqBody []byte, statusCode int, respHeaders http.Header, respBody []byte) *ValidationResult {
	result := &ValidationResult{
		Type:  "response",
		Valid: true,
	}

	if !p.config.ValidateResponse {
		return result
	}

	// Build full interaction for response validation
	interaction := &Interaction{
		Method:          r.Method,
		Path:            p.buildPath(r),
		Headers:         p.extractHeaders(r.Header),
		Body:            string(reqBody),
		StatusCode:      statusCode,
		ResponseHeaders: p.extractHeaders(respHeaders),
		ResponseBody:    string(respBody),
	}

	validationResult, err := p.validateInteraction(ctx, interaction)
	if err != nil {
		result.Valid = false
		result.Errors = []string{err.Error()}
		return result
	}

	result.Valid = validationResult.Valid
	result.Errors = validationResult.Errors

	return result
}

// validateInteraction performs the actual validation using configured backend.
func (p *Producer) validateInteraction(ctx context.Context, interaction *Interaction) (*ValidationResult, error) {
	p.mu.RLock()
	defer p.mu.RUnlock()

	return p.config.Validator.Validate(ctx, p.config.SchemaID, interaction)
}

// buildPath constructs the full path including query string.
func (p *Producer) buildPath(r *http.Request) string {
	path := r.URL.Path
	if r.URL.RawQuery != "" {
		path += "?" + r.URL.RawQuery
	}
	return path
}

// extractHeaders converts http.Header to a simple map.
func (p *Producer) extractHeaders(h http.Header) map[string]string {
	headers := make(map[string]string)
	for k, v := range h {
		if len(v) > 0 {
			headers[k] = v[0]
		}
	}
	return headers
}

// HandleRequestValidationFailure processes a request validation failure based on mode.
// Returns true if the request should continue, false if it was rejected.
func (p *Producer) HandleRequestValidationFailure(w http.ResponseWriter, r *http.Request, result *ValidationResult) bool {
	// Call custom handler if configured
	if p.config.OnRequestFailure != nil {
		return p.config.OnRequestFailure(w, r, result)
	}

	switch p.config.Mode {
	case ModeStrict:
		// Reject with 400 Bad Request
		p.writeErrorResponse(w, result)
		return false

	case ModeWarn:
		// Log and continue
		p.logValidationFailure("request", r, result)
		return true

	case ModeShadow:
		// Metrics only, continue
		recordValidationMetrics("request", result)
		return true
	}

	return true
}

// HandleResponseValidationFailure processes a response validation failure based on mode.
func (p *Producer) HandleResponseValidationFailure(r *http.Request, result *ValidationResult) {
	// Call custom handler if configured
	if p.config.OnResponseFailure != nil {
		p.config.OnResponseFailure(r, result)
		return
	}

	switch p.config.Mode {
	case ModeStrict, ModeWarn:
		// Log the failure (can't modify response - already sent)
		p.logValidationFailure("response", r, result)

	case ModeShadow:
		// Metrics only
		recordValidationMetrics("response", result)
	}
}

// writeErrorResponse writes a standardized error response for validation failures.
func (p *Producer) writeErrorResponse(w http.ResponseWriter, result *ValidationResult) {
	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(http.StatusBadRequest)

	response := map[string]any{
		"error":   "Request validation failed",
		"details": result.Errors,
	}

	_ = json.NewEncoder(w).Encode(response)
}

// logValidationFailure logs a validation failure.
func (p *Producer) logValidationFailure(validationType string, r *http.Request, result *ValidationResult) {
	log.Printf("[CVT] %s validation failed for %s %s: %v",
		validationType, r.Method, r.URL.Path, result.Errors)
}

// ResponseCapture wraps http.ResponseWriter to capture the response.
type ResponseCapture struct {
	http.ResponseWriter
	StatusCode int
	Body       bytes.Buffer
	written    bool
}

// NewResponseCapture creates a new ResponseCapture wrapping the given ResponseWriter.
func NewResponseCapture(w http.ResponseWriter) *ResponseCapture {
	return &ResponseCapture{
		ResponseWriter: w,
		StatusCode:     http.StatusOK,
	}
}

// WriteHeader captures the status code and writes it to the underlying writer.
func (rc *ResponseCapture) WriteHeader(code int) {
	if !rc.written {
		rc.StatusCode = code
		rc.ResponseWriter.WriteHeader(code)
		rc.written = true
	}
}

// Write captures the body and writes it to the underlying writer.
func (rc *ResponseCapture) Write(b []byte) (int, error) {
	if !rc.written {
		rc.WriteHeader(http.StatusOK)
	}
	rc.Body.Write(b)
	return rc.ResponseWriter.Write(b)
}

// CaptureRequestBody reads and restores the request body.
func CaptureRequestBody(r *http.Request) ([]byte, error) {
	if r.Body == nil {
		return nil, nil
	}

	body, err := io.ReadAll(r.Body)
	if err != nil {
		return nil, err
	}

	// Restore the body for the handler
	r.Body = io.NopCloser(bytes.NewReader(body))

	return body, nil
}

File: sdks/java/src/main/java/io/github/sahina/sdk/producer/Producer.java
package io.github.sahina.sdk.producer;

import java.util.Map;
import java.util.concurrent.atomic.AtomicLong;
import java.util.logging.Level;
import java.util.logging.Logger;

/**
 * Producer handles server-side validation of HTTP requests and responses.
 *
 * <p>Example:
 * <pre>{@code
 * ProducerConfig config = ProducerConfig.builder()
 *         .schemaId("my-api")
 *         .validator(myValidator)
 *         .mode(ValidationMode.STRICT)
 *         .build();
 *
 * Producer producer = new Producer(config);
 *
 * // Validate request
 * ProducerValidationResult result = producer.validateRequest(
 *         "GET", "/users", headers, null);
 * if (!result.isValid()) {
 *     // Handle failure
 * }
 * }</pre>
 */
public class Producer {
    private static final Logger LOGGER = Logger.getLogger(Producer.class.getName());

    private final ProducerConfig config;

    // Metrics
    private static final AtomicLong requestValidations = new AtomicLong(0);
    private static final AtomicLong requestValidationsPassed = new AtomicLong(0);
    private static final AtomicLong requestValidationsFailed = new AtomicLong(0);
    private static final AtomicLong responseValidations = new AtomicLong(0);
    private static final AtomicLong responseValidationsPassed = new AtomicLong(0);
    private static final AtomicLong responseValidationsFailed = new AtomicLong(0);
    private static final AtomicLong requestsRejected = new AtomicLong(0);

    /**
     * Creates a new Producer with the given configuration.
     *
     * @param config The producer configuration
     */
    public Producer(ProducerConfig config) {
        this.config = config;
    }

    /**
     * Check if a path should be validated.
     *
     * @param path The path to check
     * @return true if the path should be validated
     */
    public boolean shouldValidatePath(String path) {
        return config.shouldValidatePath(path);
    }

    /**
     * Validates an incoming HTTP request.
     *
     * @param method  HTTP method (GET, POST, etc.)
     * @param path    Request path including query string
     * @param headers Request headers
     * @param body    Request body (as string)
     * @return Validation result
     */
    public ProducerValidationResult validateRequest(
            String method,
            String path,
            Map<String, String> headers,
            String body) {

        if (!config.isValidateRequest()) {
            return ProducerValidationResult.success("request");
        }

        try {
            ProducerValidationResult result = config.getValidator().validate(
                    config.getSchemaId(),
                    method,
                    path,
                    headers,
                    body,
                    200, // Minimal valid response for request-only validation
                    null,
                    "{}");

            return new ProducerValidationResult(result.isValid(), result.getErrors(), "request");
        } catch (Exception e) {
            LOGGER.log(Level.SEVERE, "[" + config.getLogPrefix() + "] Request validation error", e);
            return ProducerValidationResult.success("request"); // Continue on error
        }
    }

    /**
     * Validates an outgoing HTTP response.
     *
     * @param method          HTTP method of the original request
     * @param path            Request path including query string
     * @param requestHeaders  Original request headers
     * @param requestBody     Original request body
     * @param statusCode      Response status code
     * @param responseHeaders Response headers
     * @param responseBody    Response body (as string)
     * @return Validation result
     */
    public ProducerValidationResult validateResponse(
            String method,
            String path,
            Map<String, String> requestHeaders,
            String requestBody,
            int statusCode,
            Map<String, String> responseHeaders,
            String responseBody) {

        if (!config.isValidateResponse()) {
            return ProducerValidationResult.success("response");
        }

        try {
            ProducerValidationResult result = config.getValidator().validate(
                    config.getSchemaId(),
                    method,
                    path,
                    requestHeaders,
                    requestBody,
                    statusCode,
                    responseHeaders,
                    responseBody);

            return new ProducerValidationResult(result.isValid(), result.getErrors(), "response");
        } catch (Exception e) {
            LOGGER.log(Level.SEVERE, "[" + config.getLogPrefix() + "] Response validation error", e);
            return ProducerValidationResult.success("response"); // Continue on error
        }
    }

    /**
     * Handles request validation failure based on mode.
     *
     * @param result  The validation result
     * @param request The original request object (framework-specific)
     * @return Object array: [shouldContinue (Boolean), customResponse (Object or null)]
     */
    public Object[] handleRequestFailure(ProducerValidationResult result, Object request) {
        // Call custom handler if configured
        if (config.getOnRequestFailure() != null) {
            try {
                Object customResponse = config.getOnRequestFailure().apply(result, request);
                if (customResponse != null) {
                    return new Object[]{false, customResponse};
                }
            } catch (Exception e) {
                LOGGER.log(Level.WARNING, "[" + config.getLogPrefix() + "] Error in onRequestFailure handler", e);
            }
        }

        switch (config.getMode()) {
            case STRICT:
                recordRejection();
                return new Object[]{false, null};

            case WARN:
                logValidationFailure("request", request, result);
                return new Object[]{true, null};

            case SHADOW:
                recordValidationMetrics("request", result);
                return new Object[]{true, null};

            default:
                return new Object[]{true, null};
        }
    }

    /**
     * Handles response validation failure based on mode.
     *
     * @param result   The validation result
     * @param request  The original request object
     * @param response The response object
     */
    public void handleResponseFailure(ProducerValidationResult result, Object request, Object response) {
        // Call custom handler if configured
        if (config.getOnResponseFailure() != null) {
            try {
                config.getOnResponseFailure().accept(result);
                return;
            } catch (Exception e) {
                LOGGER.log(Level.WARNING, "[" + config.getLogPrefix() + "] Error in onResponseFailure handler", e);
            }
        }

        switch (config.getMode()) {
            case STRICT:
            case WARN:
                // Log the failure (can't modify response - already sent)
                logValidationFailure("response", request, result);
                break;

            case SHADOW:
                // Metrics only
                recordValidationMetrics("response", result);
                break;
        }
    }

    /**
     * @return The validation mode
     */
    public ValidationMode getMode() {
        return config.getMode();
    }

    private void logValidationFailure(String type, Object request, ProducerValidationResult result) {
        LOGGER.log(Level.WARNING,
                String.format("[%s] %s validation failed: %s",
                        config.getLogPrefix(), type, result.getErrors()));
    }

    // Static metrics methods

    /**
     * Records validation metrics.
     *
     * @param type   Type of validation ("request" or "response")
     * @param result The validation result
     */
    public static void recordValidationMetrics(String type, ProducerValidationResult result) {
        if ("request".equals(type)) {
            requestValidations.incrementAndGet();
            if (result.isValid()) {
                requestValidationsPassed.incrementAndGet();
            } else {
                requestValidationsFailed.incrementAndGet();
            }
        } else {
            responseValidations.incrementAndGet();
            if (result.isValid()) {
                responseValidationsPassed.incrementAndGet();
            } else {
                responseValidationsFailed.incrementAndGet();
            }
        }
    }

    /**
     * Records a request rejection.
     */
    public static void recordRejection() {
        requestsRejected.incrementAndGet();
    }

    /**
     * Gets the current metrics snapshot.
     *
     * @return Metrics snapshot
     */
    public static Metrics getMetrics() {
        return new Metrics(
                requestValidations.get(),
                requestValidationsPassed.get(),
                requestValidationsFailed.get(),
                responseValidations.get(),
                responseValidationsPassed.get(),
                responseValidationsFailed.get(),
                requestsRejected.get());
    }

    /**
     * Resets all metrics.
     */
    public static void resetMetrics() {
        requestValidations.set(0);
        requestValidationsPassed.set(0);
        requestValidationsFailed.set(0);
        responseValidations.set(0);
        responseValidationsPassed.set(0);
        responseValidationsFailed.set(0);
        requestsRejected.set(0);
    }

    /**
     * Metrics snapshot.
     */
    public static class Metrics {
        private final long requestValidations;
        private final long requestValidationsPassed;
        private final long requestValidationsFailed;
        private final long responseValidations;
        private final long responseValidationsPassed;
        private final long responseValidationsFailed;
        private final long requestsRejected;

        public Metrics(long requestValidations, long requestValidationsPassed,
                       long requestValidationsFailed, long responseValidations,
                       long responseValidationsPassed, long responseValidationsFailed,
                       long requestsRejected) {
            this.requestValidations = requestValidations;
            this.requestValidationsPassed = requestValidationsPassed;
            this.requestValidationsFailed = requestValidationsFailed;
            this.responseValidations = responseValidations;
            this.responseValidationsPassed = responseValidationsPassed;
            this.responseValidationsFailed = responseValidationsFailed;
            this.requestsRejected = requestsRejected;
        }

        public long getRequestValidations() { return requestValidations; }
        public long getRequestValidationsPassed() { return requestValidationsPassed; }
        public long getRequestValidationsFailed() { return requestValidationsFailed; }
        public long getResponseValidations() { return responseValidations; }
        public long getResponseValidationsPassed() { return responseValidationsPassed; }
        public long getResponseValidationsFailed() { return responseValidationsFailed; }
        public long getRequestsRejected() { return requestsRejected; }
    }
}
